tests/ui/option_string.stderr

```diff
@@ -0,0 +1,5 @@
+error: unsupported element type of Option
+ --> tests/ui/option_string.rs:4:19
+  |
+4 |         fn f() -> Option<String>;
+  |                   ^^^^^^^^^^^^^^
```

Comment:
The concern is that the currently supported types are all relying on the fact that they are:
1) Equal to the size of a pointer
2) Guarantee the niche optimization with a niche of all 0's: https://rust-lang.github.io/unsafe-code-guidelines/layout/enums.html#discriminant-elision-on-option-like-enums

1) I don't think this is required and maybe I should change the code to not assume this. Instead of comparing to usize I can compare to [u8; size_of::<T>()] (somehow) and check for all 0
2) is definitely required though, and while String and Vec do provide the niche optimization today, I can't find anything that actually guarantees that... The link above specifically says only references, NonZero*, and NonNull ptrs. Theoretically, some sort of small string optimization or an equivalent for Vec could break that for example. Seems extremely unlikely though.

Reply:
This might show the issue: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ef463b9e3b9c78f97a27f5dfded0b91f there are bit patterns that are not all 0 that can correspond to None, and transmuting `None` to bytes is also not all 0s it seems (surprising to me tbh, you may have to run it multiple times because it is non-deterministic). It's possible that someone could be sufficiently clever to work around this though :p

Reply:
I now realize that `Option<&Vec<T>>` and `Option<&String>` also don't work because they need to be handled specially :/ I'll see if I can support them, initial tries seem a bit tricky